### PR TITLE
op-challenger: Asterisc Support

### PR DESCRIPTION
**Pull request by pcw109550**
TL; DR: Implemented op-challenger for asterisc. Only unit test is passing. E2E tests will be implemented at asterisc repo. Need to work on challenger abstraction to avoid code duplication.

**Description**

op-challenger provides a wrapper with FPVM binary, and it only supported cannon to play dispute game, providing traces. This PR adds a wrapper with yet another FPVM, [asterisc](https://github.com/ethereum-optimism/asterisc).

Although code diff is quite large, they do not add fancy ideas. They are almost same with cannon's structure. I mirrored the structure of cannon's [trace directory](https://github.com/ethereum-optimism/optimism/tree/f8da0f3df4d0ea63ea4ed68370aed69400e92501/op-challenger/game/fault/trace/cannon). There are few building blocks to build the challenger for asterisc. I will list up the components and their needs.
- `asterisc_state.go`, `asterisc_memory.go`, `asterisc_page.go` : It reads the asterisc state, and calculate witness or statehash. Cannot be shared with cannon because they have different [state](https://github.com/ethereum-optimism/specs/blob/main/specs/experimental/fault-proof/cannon-fault-proof-vm.md#state) formats.
   - Most of the codes are copied from the asterisc repo to avoid importing asterisc.
   - Asterisc repo already imports the monorepo, so if we import asterisc here again, we get a circular dependency.
   - I temporarily copied the partial code of asterisc repo here. **Space for abstraction**: Please refer to next paragraph.
- `prestate.go`: It reads the prestate using `asterisc_state.go`. Mirroring `cannon_state.go`.
- `executor.go`: Wrapper of asterisc binary, supplying flags.
- `provider.go`: Trace provider for asterisc.
- `preimage.go`, `local.go`: These were imported from cannon because these are not cannon specific. No new files added for these.

Asterisc specific flags where added to op-challenger. Copied cannon flags and configs, but changed their name to contain asterisc instead of cannon. Tweaked op-challenger to support asterisc game type = `2`.

dispute-mon will now accept asterisc game type as valid.

**Space for abstraction**

op-challenger must import the state/memory/page related code to interact. For example, lets compare below code segments:
At `trace/cannon/prestate.go`:
```go
func (p *CannonPrestateProvider) AbsolutePreStateCommitment(_ context.Context) (common.Hash, error) {
	...
	hash, err := mipsevm.StateWitness(state).StateHash()
	...
}
```
At `trace/asterisc/prestate.go`:
```go
func (p *AsteriscPreStateProvider) AbsolutePreStateCommitment(_ context.Context) (common.Hash, error) {
	...
	hash, err := StateWitness(state).StateHash()
	...
}
```
Each code segment requires to compute state hash from the state. For cannon, this job is relatively simple because cannon is directly embedded in the monorepo. For asterisc, different story. We cannot import asterisc to the monorepo because of circular dependency. Therefore we need to duplicate the codes, as seen in this PR.

We need better abstraction to hide these details, and this can be redesigned in the follow up PR.

**Tests**

Unit tests for asterisc tracer is added. Basic flag sanity checking unit test is added.

There are no e2e tests added! These will be added to [asterisc](https://github.com/ethereum-optimism/asterisc). Will refer to [output_cannon_test.go](https://github.com/ethereum-optimism/optimism/blob/develop/op-e2e/faultproofs/output_cannon_test.go).

To add more details about why e2e is implemented at asterisc, asterisc is expected to be maintained at separate repository. Asterisc pulls the monorepo(here!)

### CODE REVIEW

**Comment by ImTei** [comment]
What I understand is that the only thing we need to make is `TraceProvider` for Asterisc. So we don't have to follow how the `CannonTraceProvider` implements methods of `TraceProvider` like  `GetStepData()`. That means we just can make Asterisc binary return the state hash with state and proof, other than the challenger calculates the hash by itself. Then we can remove all duplicated code for Asterisc state hash calculation - such as memory implementation. (or we can also leave a minimal type definition) Let's figure out what we should change!

File: op-challenger/game/fault/trace/asterisc/asterisc_state.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+	// LastHint is optional metadata, and not part of the VM state itself.
+	// It is used to remember the last pre-image hint,
+	// so a VM can start from any state without fetching prior pre-images,
+	// and instead just repeat the last hint on setup,
+	// to make sure pre-image requests can be served.
+	// The first 4 bytes are a uin32 length prefix.
+	// Warning: the hint MAY NOT BE COMPLETE. I.e. this is buffered,
+	// and should only be read when len(LastHint) > 4 && uint32(LastHint[:4]) >= len(LastHint[4:])
+	LastHint hexutil.Bytes `json:"lastHint,omitempty"`
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/asterisc_state.go:46]: We also need modifications for asterisc repo then

File: op-challenger/game/fault/trace/asterisc/asterisc_state_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		_, err = writer.Write(testState)
+		require.NoError(t, err)
+		require.NoError(t, writer.Close())
+
+		state, err := parseState(path)
+		require.NoError(t, err)
+
+		var expected VMState
+		require.NoError(t, json.Unmarshal(testState, &expected))
+		require.Equal(t, &expected, state)
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/asterisc_state_test.go:47]: Already automatically cleaned.

File: op-challenger/game/fault/trace/asterisc/asterisc_state_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		dir := t.TempDir()
+		path := filepath.Join(dir, "state.json")
+		require.NoError(t, os.WriteFile(path, testState, 0644))
+
+		state, err := parseState(path)
+		require.NoError(t, err)
+
+		var expected VMState
+		require.NoError(t, json.Unmarshal(testState, &expected))
+		require.Equal(t, &expected, state)
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/asterisc_state_test.go:28]: `TempDir` method will fail before returning.

File: op-challenger/game/fault/trace/asterisc/prestate_test.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+func TestAbsolutePreStateCommitment(t *testing.T) {
+	dataDir := t.TempDir()
+
+	prestate := "state.json"
+
+	t.Run("StateUnavailable", func(t *testing.T) {
+		provider := newAsteriscPrestateProvider("/dir/does/not/exist", prestate)
+		_, err := provider.AbsolutePreStateCommitment(context.Background())
+		require.ErrorIs(t, err, os.ErrNotExist)
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/prestate_test.go:27]: Error it self already contains plenty info. Also msg should be the opposite.

File: op-challenger/game/fault/trace/asterisc/asterisc_memory.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	m.pages = make(map[uint64]*CachedPage)
+	m.lastPageKeys = [2]uint64{^uint64(0), ^uint64(0)}
+	m.lastPage = [2]*CachedPage{nil, nil}
+	for i, p := range pages {
+		if _, ok := m.pages[p.Index]; ok {
+			return fmt.Errorf("cannot load duplicate page, entry %d, page index %d", i, p.Index)
+		}
+		m.AllocPage(p.Index).Data = p.Data
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/asterisc_memory.go:87]: Good point. Will address this on asterisc repo.

File: op-challenger/game/fault/trace/asterisc/asterisc_memory.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		return zeroHashes[64-5+1-l]
+	}
+	if n != nil {
+		return *n
+	}
+	left := m.MerkleizeSubtree(gindex << 1)
+	right := m.MerkleizeSubtree((gindex << 1) | 1)
+	r := HashPair(left, right)
+	m.nodes[gindex] = &r
+	return r
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/asterisc_memory.go:136]: Good point. Will address this on asterisc repo.

File: op-challenger/game/fault/trace/asterisc/executor.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+		if index > bestSnap && index < traceIndex {
+			bestSnap = index
+		}
+	}
+	if bestSnap == 0 {
+		return absolutePreState, nil
+	}
+	startFrom := fmt.Sprintf("%v/%v.json.gz", snapDir, bestSnap)
+
+	return startFrom, nil
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/executor.go:197]: Yes here is definitely a optimization space, such as caching the snapshot directory filenames. We should bench if this method consumes significant time. If we really need optimization based on the benchmark, I'll go for it.

File: op-challenger/game/fault/trace/asterisc/prestate.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+var _ types.PrestateProvider = (*AsteriscPreStateProvider)(nil)
+
+type AsteriscPreStateProvider struct {
+	prestate string
+
+	prestateCommitment common.Hash
+}
+
+func NewPrestateProvider(prestate string) *AsteriscPreStateProvider {
+	return &AsteriscPreStateProvider{prestate: prestate}
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/prestate.go:20]: Good, but I want to minimize the diff here.

File: op-challenger/game/fault/trace/asterisc/provider_test.go [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+	if e.proof != nil {
+		proofFile := filepath.Join(dir, proofsDir, fmt.Sprintf("%d.json.gz", i))
+		data, err := json.Marshal(e.proof)
+		if err != nil {
+			return err
+		}
+		return writeGzip(proofFile, data)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/provider_test.go:270]: Good, but I want to minimize the diff here.

File: op-challenger/game/fault/trace/asterisc/executor.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+		return fmt.Errorf("could not create preimage cache directory %v: %w", dataDir, err)
+	}
+	if err := os.MkdirAll(proofDir, 0755); err != nil {
+		return fmt.Errorf("could not create proofs directory %v: %w", proofDir, err)
+	}
+	e.logger.Info("Generating trace", "proof", end, "cmd", e.asterisc, "args", strings.Join(args, ", "))
+	execStart := time.Now()
+	err = e.cmdExecutor(ctx, e.logger.New("proof", end), e.asterisc, args...)
+	e.metrics.RecordAsteriscExecutionTime(time.Since(execStart).Seconds())
+	return err
```
[the remark this answers is not in the document]
Comment (pcw109550) [reply op-challenger/game/fault/trace/asterisc/executor.go:142]: We must record the command execution time even though cmd failed.

**Comment by pcw109550** [comment]
Closing this PR; Refactored implementation is at https://github.com/ethereum-optimism/optimism/pull/10094.
